[visionOS] Crash in com.apple.WebKit.GPU at AVFCore:  -[AVSampleBufferVideoRenderer enqueueSampleBuffer:]

jyavenard opened the pull request:
#### e6810585bf81825a64e033621812a68e36888d28
<pre>
[visionOS] Crash in com.apple.WebKit.GPU at AVFCore:  -[AVSampleBufferVideoRenderer enqueueSampleBuffer:]
<a href="https://bugs.webkit.org/show_bug.cgi?id=288234">https://bugs.webkit.org/show_bug.cgi?id=288234</a>
<a href="https://rdar.apple.com/145047791">rdar://145047791</a>

Reviewed by NOBODY (OOPS!).

It was possible that we would enqueue a frame to the AVSampleBufferRenderer
after all its video targets had been removed, causing it to throw.

We move the removal of the video targets to the `VideoMediaSampleRenderer`
destructor to reduce the potential number of concurrent access to the
AVSampleBufferRenderer.
Additionally, we endure that the TimerDispatchSource has been cancelled
first to further guarantee that no frames can be enqueued on a shutting down
AVSampleBufferRender.

Fly-By: It was possible for the VideoMediaSampleRenderer::m_decompressionSession
member to be cleared without thread synchronisation. We now ensure that the
read operation is guarded by a lock. It practice it doesn&apos;t make a difference
as the m_decompressionSession was only ever cleared in the destructor.

* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::destroyVideoRenderer):
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h: Make class final as destructor isn&apos;t virtual.
(WebCore::VideoMediaSampleRenderer::create): Deleted.
(WebCore::VideoMediaSampleRenderer::prefersDecompressionSession const): Deleted.
(WebCore::VideoMediaSampleRenderer::as const): Deleted.
(WebCore::VideoMediaSampleRenderer::WTF_GUARDED_BY_CAPABILITY): Deleted.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::~VideoMediaSampleRenderer):
(WebCore::VideoMediaSampleRenderer::decompressionSession const):
(WebCore::VideoMediaSampleRenderer::stopRequestingMediaData):
(WebCore::VideoMediaSampleRenderer::prefersDecompressionSession const):
(WebCore::VideoMediaSampleRenderer::setPrefersDecompressionSession):
(WebCore::VideoMediaSampleRenderer::enqueueSample):
(WebCore::VideoMediaSampleRenderer::decodeNextSample):
(WebCore::VideoMediaSampleRenderer::initializeDecompressionSession):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueueAndDisplay):
(WebCore::VideoMediaSampleRenderer::flush):
(WebCore::VideoMediaSampleRenderer::resetReadyForMoreSample):
(WebCore::VideoMediaSampleRenderer::expectMinimumUpcomingSampleBufferPresentationTime):
(WebCore::VideoMediaSampleRenderer::resetUpcomingSampleBufferPresentationTimeExpectations):
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer):
(WebCore::VideoMediaSampleRenderer::totalVideoFrames const):
(WebCore::VideoMediaSampleRenderer::droppedVideoFrames const):
(WebCore::VideoMediaSampleRenderer::corruptedVideoFrames const):
(WebCore::VideoMediaSampleRenderer::totalFrameDelay const):
(WebCore::VideoMediaSampleRenderer::isUsingDecompressionSession const): Deleted.
</pre>

aestes commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -61,6 +61,12 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
+@interface AVSampleBufferVideoRenderer (Staging_127455709)
+- (void)removeAllVideoTargets;
+@end
+#endif
```
This staging should no longer be necessary since the SPI has been available in SDKs for a while at this point; instead we should just declare this in AVFoundationSPI.h for the sake of Public SDK builds.

jyavenard replied:
I'll be taking this in a follow up as here I'm really just moving code, and this should be going to current release

aestes commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -157,19 +163,31 @@ static CFComparisonResult compareBuffers(CMBufferRef buf1, CMBufferRef buf2, voi
         [renderer flush];
         [renderer stopRequestingMediaData];
+#if ENABLE(LINEAR_MEDIA_PLAYER)
+        if ([renderer respondsToSelector:@selector(removeAllVideoTargets)])
```
This `-respondsToSelector:` check should no longer be necessary.

jyavenard committed:
[visionOS] Crash in com.apple.WebKit.GPU at AVFCore:  -[AVSampleBufferVideoRenderer enqueueSampleBuffer:]
https://bugs.webkit.org/show_bug.cgi?id=288234
rdar://145047791

Reviewed by Andy Estes.

It was possible that we would enqueue a frame to the AVSampleBufferRenderer
after all its video targets had been removed, causing it to throw.

We move the removal of the video targets to the `VideoMediaSampleRenderer`
destructor to reduce the potential number of concurrent access to the
AVSampleBufferRenderer.
Additionally, we endure that the TimerDispatchSource has been cancelled
first to further guarantee that no frames can be enqueued on a shutting down
AVSampleBufferRender.

Fly-By: It was possible for the VideoMediaSampleRenderer::m_decompressionSession
member to be cleared without thread synchronisation. We now ensure that the
read operation is guarded by a lock. It practice it doesn't make a difference
as the m_decompressionSession was only ever cleared in the destructor.

* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::destroyVideoRenderer):
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h: Make class final as destructor isn't virtual.
(WebCore::VideoMediaSampleRenderer::create): Deleted.
(WebCore::VideoMediaSampleRenderer::prefersDecompressionSession const): Deleted.
(WebCore::VideoMediaSampleRenderer::as const): Deleted.
(WebCore::VideoMediaSampleRenderer::WTF_GUARDED_BY_CAPABILITY): Deleted.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::~VideoMediaSampleRenderer):
(WebCore::VideoMediaSampleRenderer::decompressionSession const):
(WebCore::VideoMediaSampleRenderer::stopRequestingMediaData):
(WebCore::VideoMediaSampleRenderer::prefersDecompressionSession const):
(WebCore::VideoMediaSampleRenderer::setPrefersDecompressionSession):
(WebCore::VideoMediaSampleRenderer::enqueueSample):
(WebCore::VideoMediaSampleRenderer::decodeNextSample):
(WebCore::VideoMediaSampleRenderer::initializeDecompressionSession):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueueAndDisplay):
(WebCore::VideoMediaSampleRenderer::flush):
(WebCore::VideoMediaSampleRenderer::resetReadyForMoreSample):
(WebCore::VideoMediaSampleRenderer::expectMinimumUpcomingSampleBufferPresentationTime):
(WebCore::VideoMediaSampleRenderer::resetUpcomingSampleBufferPresentationTimeExpectations):
(WebCore::VideoMediaSampleRenderer::copyDisplayedPixelBuffer):
(WebCore::VideoMediaSampleRenderer::totalVideoFrames const):
(WebCore::VideoMediaSampleRenderer::droppedVideoFrames const):
(WebCore::VideoMediaSampleRenderer::corruptedVideoFrames const):
(WebCore::VideoMediaSampleRenderer::totalFrameDelay const):
(WebCore::VideoMediaSampleRenderer::isUsingDecompressionSession const): Deleted.

Canonical link: https://commits.webkit.org/291014@main
